### PR TITLE
T13557: add support for booting from an ISO image

### PR DESCRIPTION
**Pull request by muresandone**
https://phabricator.endlessm.com/T13557

### CODE REVIEW

File: dracut/image-boot/eos-image-boot-setup [hunk 1]
```diff
@@ -18,12 +18,30 @@ if getargbool 0 endless.live_boot; then
...
+
+    # mount the iso image
+    mkdir -p /cd
+    mount ${host_device} /cd
+
+    # manually create the loop device needed for the device mapper
+    mknod -m660 $loopdev b 7 8
+
+    # setup the image partitions
+    losetup -v -r $loopdev "/cd/${image_path}"
```
Comment (wjt) [inline dracut/image-boot/eos-image-boot-setup]: Might be able to pass `--partscan` here and avoid needing a separate `kpartx` call?

File: dracut/image-boot/eos-image-boot-setup [hunk 2]
```diff
@@ -18,12 +18,30 @@ if getargbool 0 endless.live_boot; then
...
+    kpartx -r -a -v $loopdev
+
+    # if not an iso: fallback to normal booting
```
Comment (wjt) [inline dracut/image-boot/eos-image-boot-setup]: This comment seems misleading to me, it doesn't describe what `exit 0` does here. Maybe more like “for ISOs, that's it: no extent-map needed.”

File: dracut/image-boot/eos-image-boot-setup [hunk 3]
```diff
@@ -18,12 +18,30 @@ if getargbool 0 endless.live_boot; then
...
 
-fstype=$(lsblk --noheadings -o FSTYPE "${host_device}")
+fstype=$(lsblk -d --noheadings -o FSTYPE "${host_device}")
 if [ $? != 0 ]; then
   echo "image-boot: failed to detect filesystem type"
   exit 1
 fi
 
+if [ "$fstype" = "iso9660" ]; then
+    loopdev="/dev/mapper/endless-image"
```
Comment (wjt) [inline dracut/image-boot/eos-image-boot-setup]: Can we explain what we're doing in this block? “For ISO host filesystems, we can use a normal loop device, but its path needs to match the device created by the dm-setup logic on other filesystems.” perhaps?

File: dracut/image-boot/eos-image-boot-setup (hunk 1 above)
Comment (dsd) [inline dracut/image-boot/eos-image-boot-setup]: I recommend avoiding that, at least I found it to be racy a few years ago. https://lkml.org/lkml/2012/9/7/535

File: dracut/image-boot/eos-image-boot-setup (hunk 3 above)
Comment (dsd) [inline dracut/image-boot/eos-image-boot-setup]: I'm also curious why we can't just let losetup pick a loop device of its own name and number

**Comment by muresandone** [comment]
Commit updates, please review.

**Comment by dsd** [comment]
I don't think it would be unreasonable to improve the logic in those places
(if relevant) to allow for ISO support - but I'll leave this decision to
Will.

**Comment by dsd** [comment]
Or alternatively we could create the loopback device then create a
device-mapper linear device that corresponds exactly to it.

File: dracut/image-boot/eos-image-boot-setup (hunk 1 above)
Comment (wjt) [inline dracut/image-boot/eos-image-boot-setup]: Interesting, thanks!

**Comment by wjt** [comment]
@dsd interesting idea. So (pseudocode) we'd have an `iso9660` branch in the fstype switch, which would do something like:

```bash
mount $host_device /cd
host_device=`losetup --find --show /cd/${image_path}`
# so $host_device == /dev/loop0 in practice
size_bytes=`lsblk --output SIZE --nodeps --bytes --noheadings ${host_device}`
extents="0 ${size_bytes}"
```

and then the rest of the script would continue as "normal", setting up a trivial mapped device corresponding to the entire loopback device, etc. I like the sound of this more than creating a loop device with the right maj/min numbers by hand, even if there's an extra level of indirection. If we were to add xz decompression to device mapper then this indirection would be needed anyway, of course!

I hear your suggestion about changing the places where `/dev/mapper/endless-image` is hard-coded. In `eos-installer`, it would be easy enough to look up the parent device for the root filesystem. For the systemd units written by `eos-image-boot-generator`, could one arrange for them to be generated by `eos-image-boot-setup` itself? Otherwise we'd have to duplicate the “is it an ISO?” logic in a number of places, which isn't necessarily an improvement. My feeling is that arranging for that well-known path to be created by dm-setup in the normal way (as above) might be enough.

**Comment by wjt** [comment]
At @carlocaione's suggestion, let's merge the Dracut fix; I'll open another PR for the image-boot patch.

**Comment by carlocaione** [comment]
@wjt trying your idea but device-mapper doesn't deal well with loopback devices. When mapping `loop0` with a table like `0 12202152 /dev/loop0 0`, DM complains: `device-mapper: table: 253:0: /dev/loop0: unknown target type`

**Comment by dsd** [comment]
Try: 0 12202152 linear /dev/loop0 0

**Comment by carlocaione** [comment]
@dsd of course. Sorry. It works. Now we have to decide if we like the extra loop level :)

**Pull request merged**